subsys/bluetooth/host/gatt.c

```diff
@@ -2542,20 +2583,146 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
 int bt_gatt_notify_multiple(struct bt_conn *conn, uint16_t num_params,
 			    struct bt_gatt_notify_params *params)
 {
-	int i, ret;
+	int i;
+	int err;
+	size_t total_len = 0;
+	struct net_buf *buf;
+	struct notify_data data;
 
 	__ASSERT(params, "invalid parameters\n");
 	__ASSERT(num_params, "invalid parameters\n");
 	__ASSERT(params->attr, "invalid parameters\n");
 
+	if (num_params == 1) {
+		/* Use the standard notification API when sending only one
+		 * notification.
+		 */
+		return -EINVAL;
+	}
+
+	if (conn == NULL) {
+		/* Use the standard notification API to send to all connected
+		 * peers.
+		 */
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_READY)) {
+		return -EAGAIN;
+	}
+
+	if (conn && conn->state != BT_CONN_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+#if defined(CONFIG_BT_GATT_ENFORCE_CHANGE_UNAWARE)
+	/* BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2350:
+	 * Except for the Handle Value indication, the  server shall not send
+	 * notifications and indications to such a client until it becomes
+	 * change-aware.
+	 */
+	if (!bt_gatt_change_aware(conn, false)) {
+		return -EAGAIN;
+	}
+#endif
+
+	/* This API guarantees an ATT_MULTIPLE_HANDLE_VALUE_NTF over the air. */
+	if (!gatt_cf_notify_multi(conn)) {
+		return -EOPNOTSUPP;
+	}
+
 	for (i = 0; i < num_params; i++) {
-		ret = bt_gatt_notify_cb(conn, &params[i]);
-		if (ret < 0) {
-			return ret;
+		total_len += params[i].len;
+	}
+
+	/* PDU length is specified with a 16-bit value. */
+	if (total_len > UINT16_MAX) {
+		return -ERANGE;
+	}
+
+	/* Check there is a bearer with a high enough MTU. */
+	if (bt_att_get_mtu(conn) <
+	    (sizeof(struct bt_att_notify_mult) + total_len)) {
+		return -ERANGE;
+	}
+
+	/* The current implementation requires the same callbacks and
+	 * user_data.
+	 */
+	for (i = 1; i < num_params; i++) {
+		if ((params[0].func != params[i].func) ||
+		    (params[0].user_data != params[i].user_data)) {
+			return -EINVAL;
 		}
 	}
 
+	/* Confirm that the connection has the correct level of security. */
+	for (i = 0; i < num_params; i++) {
+		if (bt_gatt_check_perm(conn, params[i].attr,
+				       BT_GATT_PERM_READ_ENCRYPT |
+				       BT_GATT_PERM_READ_AUTHEN)) {
+			BT_WARN("Link is not encrypted");
+			return -EPERM;
+		}
+	}
+
+	/* Send any outstanding notifications */
+	gatt_notify_flush(conn);
+
+	/* Build the PDU */
+	buf = bt_att_create_pdu(conn, BT_ATT_OP_NOTIFY_MULT,
+				sizeof(struct bt_att_notify_mult) + total_len);
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	/* Register the callback. It will be called num_params times. */
+	bt_att_set_tx_meta_data(buf, params->func, params->user_data);
+	bt_att_increment_tx_meta_data_attr_count(buf, num_params);
+
+	for (i = 0; i < num_params; i++) {
```

Comment:
We are iterating on the `params` 4 times by now; can't we reduce this to a single iteration that validates *all* input, and then a single iteration where encode the data? 

e.g. the `params[i].uuid` and `!data.handle` checks can probably be done together with the `bt_gatt_check_perm` validation.

Reply:
Yep I was a bit concerned about that, I'll do it.